src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.EventDispatch.cs

```diff
@@ -78,9 +78,18 @@ private void SetNotFoundResponse(object? sender, EventArgs args)
     {
         if (_httpContext.Response.HasStarted)
         {
-            throw new InvalidOperationException("Cannot set a NotFound response after the response has already started.");
+            // We cannot set a NotFound code after the response has already started
+            var navigationManager = _httpContext.RequestServices.GetRequiredService<NavigationManager>();
+            if (navigationManager is null)
+            {
+                throw new InvalidOperationException("The NavigationManager service is not available. Cannot navigate to a 404 page.");
+            }
+            var notFoundUri = $"{navigationManager.BaseUri}not-found";
+            navigationManager.NavigateTo(notFoundUri);
+            return;
         }
         _httpContext.Response.StatusCode = StatusCodes.Status404NotFound;
+        _httpContext.Response.ContentType = null;
```

Comment:
What's the reason behind this bit? My understanding is that we didn't do this in the past. If the request bubbles up and some other part of the stack handles the 404 I would expect that part to set the content type anyways.

Reply:
We didn't do it because we didn't set re-execution. In the middleware code we are checking if content type is empty and by default, it's not.

https://github.com/dotnet/aspnetcore/blob/55a1ae7c6de104a9dc50d9dce8e44753ae82345a/src/Middleware/Diagnostics/src/StatusCodePage/StatusCodePagesMiddleware.cs#L71

To not to edit the condition, I started cleaning this field, so that processing of re-execution won't be aborted.